Use weakref in storing tensors as keys (follow-up to #111470)

pearu opened the pull request:
This PR addresses the discussion items in https://github.com/pytorch/pytorch/pull/111470#discussion_r1369008167, that is,
- use weakref when storing tensors as keys,
- add `storage_offset` to the key data,
- and revise the description of the `TensorAsKey` utility.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112337
* __->__ #112076
* #112154

pearu committed:
Use weakref in storing tensors as keys (follow-up to #111470)

[ghstack-poisoned]

pearu committed:
Update on "Use weakref in storing tensors as keys (follow-up to #111470)"

This PR addresses the discussion items in https://github.com/pytorch/pytorch/pull/111470#discussion_r1369008167, that is,
- use weakref when storing tensors as keys,
- add `storage_offset` to the key data,
- and revise the description of the `TensorAsKey` utility.




[ghstack-poisoned]

albanD commented on torch/sparse/_triton_ops.py:
```diff
@@ -727,7 +741,10 @@ def _bsr_scatter_mm_indices_data(indices_format, M, K, N, Ms, Ks, nbatches, SPLI
         a = non_zero_row_indices * (Ms * N)
         r_offsets = (a + b).view(-1)
-        c_indices = crow_indices
+        # crow_indices consitutes a part of a key in lru_cache as well
```
I'm confused about what this means tbh.

pearu replied:
The definition of `_bsr_scatter_mm_indices_data` function is:
```python
@lru_cache(maxsize=128)
def _bsr_scatter_mm_indices_data(indices_format, M, K, N, Ms, Ks, nbatches, SPLIT_N, crow_indices_as_key, col_indices_as_key):
    ...
```
which means that a key to LRU cache constitutes of all arguments to the function:
```
lru_cache_key = (indices_format, M, K, N, Ms, Ks, nbatches, SPLIT_N, crow_indices_as_key, col_indices_as_key)
```
while the value of the LRU cache is the result of the function. Notice that the argument `crow_indices_as_key` is `TensorAsKey(crow_indices)`.
On the other hand, the result of the function is a tuple `(indices_format, c_indices, r_offsets, q_offsets)` where originally `c_indices` was `crow_indices`.

So, this means that the tensor `crow_indices` is stored both in the key as well as in the value of LRU cache. Now, when the `crow_indices` variable goes out of scope, the corresponding tensor ought to be garbage collected (assuming that `crow_indices` appears as weakref instance in the key). However, the tensor cannot be garbage collected because it is referenced by the value of the LRU cache. Therefore, the `crow_indices` tensor must be cloned so that garbage collection would be effective when the `crow_indices` variable goes out of the scope, and later, when the LRU cache will overfill, the cloned tensor will be cleaned up as well.

pearu replied:
Hmm, after thinking about it, I think the `crow_indices` tensor in the value part of the LRU cache ought to be weakref as well. Then we don't have to wait for an overfill for the tensor to be garbage collected. I'll fix it unless I missed something right now..

pearu replied:
This comment is now removed. It turns out that it is not sufficient to use crow_indices and col_indices as keys because these have shorter lifetime than the original sparse compressed tensor. Now the original tensor is wrapped with TensorAsKey to make sure that the cached item exists in the life time of the original tensor.

pearu committed:
Update on "Use weakref in storing tensors as keys (follow-up to #111470)"

This PR addresses the discussion items in https://github.com/pytorch/pytorch/pull/111470#discussion_r1369008167, that is,
- use weakref when storing tensors as keys,
- add `storage_offset` to the key data,
- and revise the description of the `TensorAsKey` utility.




[ghstack-poisoned]

pearu committed:
Update on "Use weakref in storing tensors as keys (follow-up to #111470)"

This PR addresses the discussion items in https://github.com/pytorch/pytorch/pull/111470#discussion_r1369008167, that is,
- use weakref when storing tensors as keys,
- add `storage_offset` to the key data,
- and revise the description of the `TensorAsKey` utility.




[ghstack-poisoned]

pearu committed:
Update on "Use weakref in storing tensors as keys (follow-up to #111470)"

This PR addresses the discussion items in https://github.com/pytorch/pytorch/pull/111470#discussion_r1369008167, that is,
- use weakref when storing tensors as keys,
- add `storage_offset` to the key data,
- and revise the description of the `TensorAsKey` utility.




[ghstack-poisoned]
